Caltech-IPAC/firefly
# FIREFLY-465: adding sofia analyzer and include spectra 1d extraction from FITS image data

The purpose of this PR is to add an DP analyzer that will be used for SOFIA data product purposes. See ticket here: [FIREFLY-465](https://jira.ipac.caltech.edu/browse/FIREFLY-465)
This PR has a counterpart PR for SOFIA in IFE [here](https://github.com/IPAC-SW/irsa-ife/pull/115).

The analyzer will trigger and add an extracted spectra to the report returned using a custom FITS HUD image extractor (also included) and a spectra model defined for SOFIA using VO UCDs and units as standard as possible.
Extracted and added spectra will depend on a couple of columns: instrument, product type and level.

The interesting code to look for and complete or update is here:
- The analyzer: `edu.caltech.ipac.firefly.server.dpanalyze.SofiaAnalyzer`
- The extractor: `edu.caltech.ipac.firefly.data.sofia.Sofia1DSpectraExtractor`
- The meta info added to the client dispatch in IFE to make use of the Firefly analyzer can be found here: `src/irsa/js/sofia/Sofia.jsx`

The analyzer will extract the data and add it to the report as a VOTable following a drafted spectral model. (See `edu.caltech.ipac.firefly.data.sofia.VOSpectraModel`)
The analyzer should be completed with other analysis, replacement or extraction as we go.

Caveats: 

1. The combinations to identify 1D spectra included in the FITS images are still in progress ([see confluence page](https://confluence.ipac.caltech.edu/pages/viewpage.action?spaceKey=IRSA&title=SOFIA+-+Plotting+flux+versus+wavelength
 )). 
2. Spectral model used is also in progress, ucds, units, etc. That will need more expertise/review.
3. What is absolutely left out the PR is how to deal with multiple plane (case of NAXIS3 or 'cube') , visualizing multiple spectra.

A way to test the 1D spectra is to 
- go to Firefly build here:
https://fireflydev.ipac.caltech.edu/firefly-465-sofia-analyzer/firefly/
- upload the table [here](http://web.ipac.caltech.edu.s3-us-west-2.amazonaws.com/staff/ejoliet/sofia/FORCAST-with-meta-analyzer.tbl), 
- select the second row
- noticed that the 'File Contents' now has 3rd part which correspond to the extracted data from the primary HDU and the chart represent flux and wavelength.
 
Another way to test is to 
- go to the SOFIA app build here:
https://irsakudevlb1.ipac.caltech.edu/firefly-465/applications/sofia/
- do a search around moving object 'Pallas' with level 2,3,4 checked and look for FORCAST, EXES or FLITECAM instruments on different product type and level. Using FORCAST table as an example, filter the table by product type column with `IN ('combspec','mrgspec','rspspec')`, first row highlighted should show a chart wavelength vs flux.
The menu 'File Contents' on those data product and other combination of instrument, product types and levels should show an additional part which represent the spectra extracted

For example, FLITECAM LEVEL 3 and product type 'IN ('calspec')'... etc.

If you got to this point and not satisfied with the testing, look at the if-then-else statement in the code `edu.caltech.ipac.firefly.server.dpanalyze.SofiaAnalyzer#analyzeFits` or see confluence page [here](https://confluence.ipac.caltech.edu/pages/viewpage.action?spaceKey=IRSA&title=SOFIA+-+Plotting+flux+versus+wavelength
 ).
